src/BrowseView.js

```diff
@@ -0,0 +1,394 @@
+          const index = this.selection_model.selected;
+          const bottom_edge = (index + 1) * 38 - this._adj.value;
+          const top_edge = bottom_edge - 38;
+          // If row is not visible after scroll_to, adjust
+          if (bottom_edge === 0) {
+            this._adj.value -= 38;
+          } else if (top_edge === this._adj.page_size) {
+            this._adj.value += 38;
+          }
+          this._scrolled_to = false;
+        }
+      });
+    }
+
+    selectItem(path) {
+      const index = this.#getItemIndex(path);
+      // If possible, overshoot scrolling by one row to ensure selected row is visible
+      index + 1 === this.selection_model.n_items
+        ? this._browse_list_view.scroll_to(
+            index,
+            Gtk.ListScrollFlags.NONE,
+            null,
+          )
+        : this._browse_list_view.scroll_to(
+            index + 1,
+            Gtk.ListScrollFlags.NONE,
+            null,
+          );
+      this.selection_model.selected = index;
+      this._scrolled_to = true;
+    }
+
+    #getItemIndex(path) {
+      let relative_index = 0; // Relative index of the item under its parent
+      let absolute_index = 0; // Index of the item in the entire model
+      let skip = 0; // Number of items to skip due to expanded rows
+
+      for (let i = 0; i < path.length; i++) {
+        while (relative_index < path[i]) {
+          const row = this._tree_model.get_row(absolute_index);
+          if (row.expanded) {
+            skip += row.children.get_n_items();
+          }
+          if (!skip) relative_index++; // Go to next sibling
+          else skip--;
+          absolute_index++;
+        }
+        // Check to ensure the last item is not expanded
+        if (i < path.length - 1) {
+          this._tree_model.get_row(absolute_index).expanded = true;
+          absolute_index++;
+          relative_index = 1;
+        }
+      }
+      return absolute_index;
+    }
+
+    #createBrowseSelectionModel() {
+      this._tree_model = Gtk.TreeListModel.new(
+        this.root_model,
+        false,
+        false,
+        (item) => item.children,
+      );
+      this.selection_model = Gtk.SingleSelection.new(this._tree_model);
+      this.selection_model.connect("selection-changed", () => {
+        // If selection changed to sync the sidebar, dont load_uri again
+        if (this.sync_sidebar) {
+          this.sync_sidebar = false;
+          return;
+        }
+        const uri = this.selection_model.selected_item.item.uri;
+        this._webview.load_uri(uri);
+      });
+      this._browse_list_view.model = this.selection_model;
+    }
+
+    async #loadDocs() {
+      return Promise.all([
+        this.#scanLibraries(Gio.File.new_for_path("/usr/share/doc")),
+        this.#scanLibraries(Gio.File.new_for_path("/usr/share/gtk-doc/html")),
+        this.#scanLibraries(Gio.File.new_for_path("/app/share/doc")),
+      ]).then(() => {
+        this.emit("browse-view-loaded");
```

Comment:
Converting the widgets into custom GObjects is very handy as it lets me do stuff like this, I listen for this signal in ``Sidebar.js`` and when it does get emitted, I use the ``root_model`` to create the ``flattened_model`` which lets me avoid long Promise-then chains